packages/hash/frontend/src/lib/subgraph.ts

```diff
@@ -0,0 +1,334 @@
+  subgraph: Subgraph,
+  linkTypeId: VersionedUri,
+): PersistedLinkType | undefined => {
+  const vertex = subgraph.vertices[linkTypeId];
+
+  if (!vertex) {
+    return undefined;
+  }
+
+  if (!isLinkTypeVertex(vertex)) {
+    throw new Error(`expected link type vertex but got: ${vertex.kind}`);
+  }
+
+  return vertex.inner;
+};
+
+/**
+ * Gets a `PersistedEntityType` by its `VersionedUri` from within the vertices of the subgraph. Returns `undefined` if
+ * the entity type couldn't be found.
+ *
+ * @param subgraph
+ * @param entityTypeId
+ * @throws if the vertex isn't a `EntityTypeVertex`
+ */
+export const getPersistedEntityType = (
+  subgraph: Subgraph,
+  entityTypeId: VersionedUri,
+): PersistedEntityType | undefined => {
+  const vertex = subgraph.vertices[entityTypeId];
+
+  if (!vertex) {
+    return undefined;
+  }
+
+  if (!isEntityTypeVertex(vertex)) {
+    throw new Error(`expected entity type vertex but got: ${vertex.kind}`);
+  }
+
+  return vertex.inner;
+};
+
+/** @todo - getPersistedEntity and getPersistedLink - https://app.asana.com/0/0/1203157172269853/f */
+
+/**
+ * Returns all `PersistedDataType`s within the vertices of the subgraph
+ *
+ * @param subgraph
+ */
+export const roots = (subgraph: Subgraph): Vertex[] => {
+  return subgraph.roots.map((root) => {
+    const rootVertex = subgraph.vertices[root];
+    if (!rootVertex) {
+      throw new Error(`looked in vertex set but failed to find root: ${root}`);
+    }
+
+    return rootVertex;
+  });
+};
+
+/**
+ * Returns all `PersistedDataType`s within the vertices of the subgraph
+ *
+ * @param subgraph
+ */
+export const persistedDataTypes = (subgraph: Subgraph): PersistedDataType[] => {
+  return Object.values(subgraph.vertices)
+    .filter(isDataTypeVertex)
+    .map((vertex) => vertex.inner);
+};
+
+/**
+ * Returns all `PersistedPropertyType`s within the vertices of the subgraph
+ *
+ * @param subgraph
+ */
+export const persistedPropertyTypes = (
+  subgraph: Subgraph,
+): PersistedPropertyType[] => {
+  return Object.values(subgraph.vertices)
+    .filter(isPropertyTypeVertex)
+    .map((vertex) => vertex.inner);
+};
+
+/**
+ * Returns all `PersistedLinkType`s within the vertices of the subgraph
+ *
+ * @param subgraph
+ */
+export const persistedLinkTypes = (subgraph: Subgraph): PersistedLinkType[] => {
+  return Object.values(subgraph.vertices)
+    .filter(isLinkTypeVertex)
+    .map((vertex) => vertex.inner);
+};
+
+/**
+ * Returns all `PersistedEntityType`s within the vertices of the subgraph
+ *
+ * @param subgraph
+ */
+export const persistedEntityTypes = (
+  subgraph: Subgraph,
+): PersistedEntityType[] => {
+  return Object.values(subgraph.vertices)
+    .filter(isEntityTypeVertex)
+    .map((vertex) => vertex.inner);
+};
+
+/** @todo - improve the typing of these */
+
+/**
+ * Returns all `PersistedLink`s within the vertices of the subgraph
+ *
+ * @param subgraph
+ */
+export const persistedLinks = (subgraph: Subgraph) => {
```

Comment:
Could we add a return type here? I suppose it would be something like
```suggestion
export const persistedLinks = (subgraph: Subgraph): PersistedLink[] => {
```
but seems to not be imported